storage/zynq7000-flash/qspi.c

```diff
@@ -0,0 +1,436 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Quad-SPI Controller driver
+ *
+ * Copyright 2021, 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "qspi.h"
+
+#include <errno.h>
+#include <string.h>
+#include <sys/time.h>
+#include <sys/mman.h>
+#include <sys/threads.h>
+#include <sys/platform.h>
+#include <sys/interrupt.h>
+
+#include <phoenix/arch/zynq7000.h>
+
+
+enum { cr = 0, sr, ier, idr, imr, er, dr, txd00, rxd, sicr, txth, rxth, gpio,
+	   lpbk = 0xe, txd01 = 0x20, txd10, txd11,
+	   lqspi_cr = 0x28, lqspi_sr, modid = 0x3f };
+
+
+struct {
+	volatile uint32_t *base;
+	handle_t cond;
+	handle_t inth;
+	handle_t lock;
+} qspi_common;
+
+
+static inline void qspi_dataMemoryBarrier(void)
+{
+	__asm__ volatile ("dmb");
+}
+
+
+static int qspi_irqHandler(unsigned int n, void *arg)
+{
+	*(qspi_common.base + idr) |= (1 << 2);
+
+	return 1;
+}
+
+
+static inline time_t qspi_timeMsGet(void)
+{
+	struct timeval time;
+	gettimeofday(&time, NULL);
+
+	return (time_t)((time.tv_sec * 1000LL) + (time.tv_usec / 1000));
+}
+
+
+void qspi_stop(void)
+{
+	/* Clean up RX Fifo */
+	while ((*(qspi_common.base + sr) & (1 << 4)))
+		*(qspi_common.base + rxd);
+
+	*(qspi_common.base + cr) |= (1 << 10);
+	*(qspi_common.base + er) = 0;
+
+	qspi_dataMemoryBarrier();
+}
+
+
+void qspi_start(void)
+{
+	*(qspi_common.base + rxth) = 0x1;
+	*(qspi_common.base + cr) &= ~(1 << 10);
+	*(qspi_common.base + er) = 0x1;
+
+	qspi_dataMemoryBarrier();
+}
+
+
+static unsigned int qspi_rxData(uint8_t *rxBuff, size_t size)
+{
+	const uint32_t data = *(qspi_common.base + rxd);
+
+	if (rxBuff != NULL) {
+		switch (size) {
+			case 0:
+				break;
+
+			case 1:
+				rxBuff[0] = data >> 24;
+				break;
+
+			case 2:
+				rxBuff[0] = (data >> 16) & 0xff;
+				rxBuff[1] = data >> 24;
+				break;
+
+			case 3:
+				rxBuff[0] = (data >> 8) & 0xff;
+				rxBuff[1] = (data >> 16) & 0xff;
+				rxBuff[2] = data >> 24;
+				break;
+
+			case 4:
+			default:
+				rxBuff[0] = data & 0xff;
+				rxBuff[1] = (data >> 8) & 0xff;
+				rxBuff[2] = (data >> 16) & 0xff;
+				rxBuff[3] = data >> 24;
+				break;
+		}
+	}
+
+	return (size < 4) ? size : 4;
```

Comment:
rxBuff was NULL, but we're still returning size

Reply:
https://github.com/phoenix-rtos/plo/pull/143/files#r729284214 - discussed before, probably in private conversation, marked as resolved by You.

Summing up, we receive data from fifo even if `rxBuff` is `NULL`, so the size has to be returned.